### DISCUSSION TITLE
Transfer discussions from draft to release

### DISCUSSION THREAD

**Discussion by cfcohen** [knowledgefutures/pubpub#1500] [Ideas]
_Before filing your bug report, please search existing issues to make sure a similar one doesn't already exist. If one does, please leave a comment or reaction in support of the issue rather than creating a new one._

_To fill out the report, please leave the below headers (**) in tact, and fill out each section. If a section isn't applicable, leave it blank or write n/a under it. Please do not delete any template sections._

### What went wrong, step-by-step?
1. Create a new publication.
2. Add comments to the publication as you write it (before publishing the first time).
3. Publish the publication.
4. The comments are visible in edit mode, but not on the published document.

### What did you expect to happen?

I expected that the comments would be published when the publication was published.

### What URL can we use to see the issue?

[Here](https://medieval.pubpub.org/pub/6cent-ch5/release/2?readingCollection=9946dc25) although it may be confusing now because I manually duplicated the comments that I made on the editable version onto the publish version.

### What browser/operating system were you using?

Linux/Firefox 89.0.2

===

### Have any screenshots? (Attach them by dragging the image here)

No, but I can create a test document and take screenshots if you have any trouble reproducing the bug from the description.

### Who reported this? (If not you)

I found and reported this issue.

**Comment by gabestein** [comment]
Thanks for adding this note! I'm going to move this to the forum for now, because the current behavior is intended, not a bug. Many people use discussions on drafts as a way to privately comment with authors before release, and do not want discussions to be made available upon release.

However, we know folks do often want to transfer certain discussions to releases, and we've discussed building release tooling to allow for that. We'll take it up with the team at our next sprint meeting. If you have any thoughts for how a feature like that could work, please let us know here.

**Reply by cfcohen** [reply]
Sorry about opening a bug for this.  Until you mentioned the alternative use case, I hadn't considered that it could even be considered a feature.   ;-)

In my case I'm the sole author transcribing a public domain book, proofing the OCR, adding hyperlinks, and occasionally adding my own comments as I produce the original release.   On that front, pubpub seems to be very well suited for this purpose, and I'm excited by the idea that pubpub could not only host public domain books (ala Project Gutenberg)  but also provide a place to discuss difficult passages, provide additional context, add contemporary observations, etc.  This seems fairly close to the "Frankenbook" use case.

Maybe a separate button to publish the comments on the draft would be a good solution.   In my case I was surprised by the current behavior, and had no easy solution for moving the existing comments (gratefully there were only about a dozen) which contributed to the perception that it was a "bug".   Another solution might be to say somewhere on the comment that it will only be seen by other editors, even after publishing.  Even tiny print with a link to documentation on the issue would be sufficient in my opinion, but the "publish comments" button seems like a better solution to me.

On a largely unrelated issue, I was similarly a little surprised that there doesn't seem to be a good way to "download" the annotations associated with the text.  Maybe there's code in the repo that I just haven't found yet (my Javascript isn't great and there seems to be a lot of code) but ideally, I'd have a Python library for programmatic access to the text and comments.  This could be used to check for broken links, watch for new comments, remove spam, gather statistics. etc.  I knew this issue wasn't a "bug", and I wasn't sure it really counted as an enhancement either, so the plan was to ask on the forums, and well... I'm here now anyway.  Any pointers or assistance you could provide would be appreciated, especially if there's an existing API in Javascript that returns JSON or something like that.   I can probably handle the Python parts if there is.

Finally, I've seen some commentary suggesting that even though the code is fully public, it's not actually easy/possible to get a standalone instance running.  Can you point me to whatever counts as the latest information on that issue?  There's no need to go into any more detail than someone already has, I just don't want to start experimenting while I'm missing important information.

**Reply by gabestein** [reply]
No worries at all! This description of your use-case is super helpful (and, also, seems like a very cool project). On some specifics:

> Maybe a separate button to publish the comments on the draft would be a good solution.

I think we actually have some internal tooling to transfer discussions that I can see if we can apply when you're ready to release. If you're interested, I'll ask about it on our next dev sprint. Longer term, we have been planning to add an option to publish non-archived comments as part of the release creation flow. I'm not sure exactly when we'll get to it, but if I'm reading correctly, I think that should support your use-case.

> I was similarly a little surprised that there doesn't seem to be a good way to "download" the annotations associated with the text.

Yeah, this is a good point, and something others have mentioned before. Unfortunately, there's no good programmatic way to do this now, although we've talked about providing an option to enable downloads with annotations in them. The reason we've hesitated there is there seems to be a lack of standards for how to represent them, which poses a bit of a problem. Under the hood we're using Pandoc for conversion, which doesn't really have first-class support for them, as far as I know. Overloading links or notes feels both complex to implement and unsatisfying. If you have suggestions or examples to show here, we'd love to see them.

> ideally, I'd have a Python library for programmatic access to the text and comments. This could be used to check for broken links, watch for new comments, remove spam, gather statistics. etc.

At the moment, we don't have a public API, and it's not high up on our roadmap, I'm afraid, as most of the folks we're serving aren't as sophisticated as you. We hope to get there, and have worked with folks to build some tools that we can run directly (poke around the `tools` directory for examples) to take care of certain tasks. If you have an idea for something like that that would help you, we'd be happy to build it, as it would probably be useful for others, too.

> Finally, I've seen some commentary suggesting that even though the code is fully public, it's not actually easy/possible to get a standalone instance running

Unfortunately, that's correct. In particular, we rely heavily on Firebase, which is proprietary and doesn't (quite yet) have a solid OS alternative with a compatible-enough API that we could make a switch to, although [Supabase](https://supabase.io/) is starting to look promising. We're also mostly focused on the moment on providing services for folks who don't have the time/expertise/resources to self-host. Most folks we would work with who do have the ability to self-host have good alternatives in platforms like Janeway and OJS. Still, it's an ambition to, one day, provide a self-contained docker image!

The best info we have on self-hosting can be found here: https://github.com/pubpub/pubpub/issues/1125

**Reply by oabb** [reply]
Hi team,

I'd like to "re-open" the discussion on this issue because we would like to use comments the other way around. We published a new release of a Pub and can see the comments of the previous release although the text has changed. Irritatingly we cannot see the comments in edit mode. When we start archiving the comments in the new release, they are also archived and no longer visible in the previous release. 

**What went wrong, step-by-step?**

    Create a new release. 
    Still see the comments of the previous release (but not in edit mode) 
    Delete comments in the new release
    Comments are also deleted in previous release

**What did you expect to happen?**

I expected that the comments would not be published in a new release automatically and if so, that it would be possible to archive comments only in the new release (because the comments were mainly related to text passage that are revised in the new version.

We also did not expect the comments to be archived in all releases when a archived in one (the last) release.

**What URL can we use to see the issue?**

[Here](https://oabb.pubpub.org/pub/mr9x0ptz/release/3)

**What browser/operating system were you using?**

Windows/Firefox 123.0

**Reply by gabestein** [reply]
Unfortunately, nothing has changed. The platform is working as intended and we do not plan to change this behavior at the moment, as the technical complexity of doing so would be huge.
